www/js/terasender/terasender.js

```diff
@@ -112,10 +112,13 @@ window.filesender.terasender = {
         if(!file) return null; // Nothing to do
         
         if(!file.endpoint) file.endpoint = this.transfer.authenticatedEndpoint(filesender.config.terasender_upload_endpoint.replace('{file_id}', file.id), file);
+
+        var chunkid = Math.floor(file.uploaded / filesender.config.upload_chunk_size);
```

Comment:
Please note the remark on Page 19,

  * It is vital for counter mode encryption that the 32-bit counter will never ever wrap around.

I chased your code a bit, but could not find a place where you constrain the size of the upload to the given number of uploads.  Please note that the last chunk may not be full, as that would lead to a need to send another, so the maximum number of chunks is 2^32 and that last chunk must not be full-sized.  (I am assuming a fixed chunk size.)

Reply:
Given that these chunks are always the 16-byte strings handled by a single AES iteration, the maximum size is 16 * 2^32 - 1.

Reply:
I think the chunk term is general enough that it is used in multiple different ways in different contexts, so it might be a good idea to clarify the above use. Normally FileSender sends files in chunks which by default are 4mb in size. This would allow for 2^32 chunks at 4mb a pop which is around 16tb before the chunk id wraps. 

This is also why I have used a new 96 bits of random material for the IV in each "chunk" as they are sort of like little 4mb files. Though I have it earmarked to move to the single 96 bits for the whole file.

There is a current limitation with uploads for encrypted files having to be less than 4tb in size. But the code should also check for the GCM limit in case the 4tb limit gets raised in the future. I'll fix that shortly.

Reply:
I got to your comment at the end of page, I'll limit to 16 * 2^32 - 1

Reply:
Changes such as fresh random material for each "chunk" have serious ramifications for the security of the system.  You need a mechanism to validate that it is all part of the same file.  AEAD might be another mechanism, and I could look into that, but **do not make assumptions of this kind** as they endanger the fine balance under which the cipher mode was designed.

The thing is, I am fencing off cases that are invisible to you, and that involve a rogue party aware of your code structure.  Assumptions like these (and not being able to check that the "chunks" belong to the same upload) seriously endanger protection against such parties.  And you could never tell from tests that these holes existed.  That's my responsibility, and the cause of specifying these validation rules and IV formats in a very particular way.  There can be variations, but implementation choices are the [only] driving force for security code, and such changes must not be made silently, for danger of opening security holes.